auth_brute_force/models/res_authentication_attempt.py

```diff
@@ -2,35 +2,160 @@
+    )
+    whitelisted = fields.Boolean(
+        compute="_compute_whitelisted",
+    )
+
+    @api.multi
+    @api.depends('remote')
+    def _compute_metadata(self):
+        for item in self:
+            url = GEOLOCALISATION_URL.format(item.remote)
+            try:
+                res = json.loads(urlopen(url, timeout=5).read())
+            except Exception:
+                _logger.error(
+                    "Couldn't fetch details from %s",
+                    url,
+                    exc_info=True,
+                )
+            else:
+                item.metadata = "\n".join(
+                    '%s: %s' % pair for pair in res.items())
+
+    @api.multi
+    def _compute_whitelisted(self):
+        whitelist = self._whitelist_remotes()
+        for one in self:
+            one.whitelisted = one.remote in whitelist
 
-    # Custom Section
     @api.model
-    def search_last_failed(self, remote):
-        last_ok = self.search(
-            [('result', '=', 'successfull'), ('remote', '=', remote)],
-            order='attempt_date desc', limit=1)
-        if last_ok:
-            return self.search([
-                ('remote', '=', remote),
-                ('attempt_date', '>', last_ok.attempt_date)])
-        else:
-            return self.search([('remote', '=', remote)])
+    def _trusted(self, remote, login):
+        """Checks if any the remote or remote+login are trusted.
+
+        :param str remote:
+            Remote IP from which the login attempt is taking place.
+
+        :param str login:
+            User login that is being tried.
+
+        :return bool:
+            ``True`` means it is trusted. ``False`` means that it is banned.
+        """
+        # Cannot ban without remote
+        if not remote:
+            return True
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        # Whitelisted remotes always pass
+        if remote in self._whitelist_remotes():
+            return True
+        # Check if remote is banned
+        limit = int(get_param("auth_brute_force.max_by_ip", 50))
+        domain_good = [
+            ("result", "=", "successful"),
+            ("remote", "=", remote),
+        ]
+        domain_bad = [
+            ("result", "!=", "successful"),
+            ("remote", "=", remote),
+        ]
+        last_ok = self.search(domain_good, order='create_date desc', limit=1)
+        recent_failures = self.search(
+            domain_bad + [("create_date", ">", last_ok.create_date)
+                          if last_ok else TRUE_LEAF],
+            count=True
+        )
+        if recent_failures >= limit:
+            _logger.warning(
+                "Authentication failed from remote '%s'. "
+                "The remote has been banned. "
+                "Login tried: %r.",
+                remote,
+                login,
+            )
+            return False
+        # Check if remote + login combination is banned
```

Comment:
Readability: The `# Check if remote is banned` and `# Check if remote + login combination is banned` parts share much of the code, might be worth factoring out into separate helper function.

Reply:
Done